File: Convert Sorted List to Binary Search Tree/Convert Sorted List to Binary Search Tree.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; next = null; }
 * }
 */
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public TreeNode sortedListToBST(ListNode head) {
        if(head==null) return null;
        int n=0;
        ListNode temp=head;
        while(temp!=null){
            n++;
            temp=temp.next;
        }
        Return A=ListToBST(head,n);
        return A.t;
    }
    public Return ListToBST(ListNode head, int n){
        if(n==1) return  new Return(new TreeNode(head.val), head.next);
        Return l=ListToBST(head, n/2);
        TreeNode root = new TreeNode(l.l.val);
        root.left=l.t;
        if((n-1)/2!=0){
            Return r=ListToBST(l.l.next, (n-1)/2);
            root.right=r.t;
            return new Return(root, r.l);
        }else{
            return new Return(root, l.l.next);
        }
    }
}
class Return{
    TreeNode t;
    ListNode l;
    Return(TreeNode tree, ListNode list){
        t=tree;
        l=list;
    }
}

File: Convert Sorted List to Binary Search Tree/Convert Sorted List to Binary Search Tree .py
# Definition for singly-linked list with a random pointer.
# class RandomListNode:
#     def __init__(self, x):
#         self.label = x
#         self.next = None
#         self.random = None

class Solution:
    # @param head, a RandomListNode
    # @return a RandomListNode
    def copyRandomList(self, head):
        if head==None:
            return head;
        original=head
        newhead=RandomListNode(0)
        newhead.next=head
        while original!=None:
            copy=RandomListNode(original.label)
            copy.next=original.next
            original.next=copy
            original=copy.next
        original=head
        while original!=None:
            if original.random==None:
                original=original.next.next
            else:
                original.next.random=original.random.next;
                original=original.next.next
        original=head
        newh=newhead
        while original!=None:
            newh.next=original.next
            original.next=original.next.next
            newh=newh.next
            original=original.next
        return newhead.next
            
